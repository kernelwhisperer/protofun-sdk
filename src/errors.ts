export class IndexerError extends Error {
  name = "IndexerError"
}
