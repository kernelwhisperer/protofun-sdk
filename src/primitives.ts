export type Timeframe = "Block" | "Minute" | "Hour" | "Day" | "Week"

export enum PriceUnit {
  ETH = "ETH",
  USD = "USD",
  GWEI = "Gwei",
}

/**
 * @internal
 */
export type BlockData = {
  baseFeePerGas: string // BigInt
  timestamp: string
}

export type Candle = {
  close: string
  high: string
  low: string
  open: string
  timestamp: string
}

export function isCandle(value: unknown): value is Candle {
  return typeof value === "object" && value !== null && "open" in value
}

export function isCandleArray(value: unknown[]): value is Candle[] {
  return Array.isArray(value) && value.length > 0 && isCandle(value[0])
}

export type QueryRequest = {
  /**
   * @default 1000
   */
  limit?: number
  priceUnit?: PriceUnit
  since?: string
  timeframe: Timeframe
  until?: string
}
export type QueryResult = Promise<Candle[]>
export type QueryFn = (request: QueryRequest) => QueryResult

export type SubscribeRequest = {
  onNewData: (data: Candle) => void
  /**
   * in milliseconds
   * @default 3000
   */
  pollingInterval?: number
  priceUnit?: PriceUnit
  since?: string
  timeframe: Timeframe
}

/**
 * @returns cleanup function used to unsubscribe
 */
export type SubscribeResult = () => void
export type SubscribeFn = (request: SubscribeRequest) => SubscribeResult
