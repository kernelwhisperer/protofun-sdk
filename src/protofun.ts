export * from "./errors"
export * from "./metric-declarations"
export * from "./primitives"
