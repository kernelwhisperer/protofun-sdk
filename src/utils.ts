import { MetricId, ProtocolId, QueryFn, SubscribeFn, Timeframe } from "./protofun"

export async function wait(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export const allTimeframes: Timeframe[] = ["Block", "Minute", "Hour", "Day", "Week"]

export type MetricFnsResult = {
  query: QueryFn
  subscribe: SubscribeFn
  supportedTimeframes: Timeframe[]
}

export async function loadMetricFns(
  protocolId: ProtocolId,
  metricId: MetricId,
  packagePath = "protofun/dist/metrics"
): Promise<MetricFnsResult> {
  const module = await import(`${packagePath}/${protocolId}/${metricId}`)
  const { default: query, supportedTimeframes = allTimeframes, subscribe } = module
  return { query, subscribe, supportedTimeframes }
}

export function getLowestTimeframe(supportedTimeframes: Timeframe[]) {
  return supportedTimeframes.find((x) => x !== "Block") as Timeframe
}
