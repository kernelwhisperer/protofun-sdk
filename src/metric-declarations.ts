import { PriceUnit, Timeframe } from "./primitives"
import { allTimeframes } from "./utils"

const METRIC_DECLARATIONS = {
  aave: [],
  btc: ["btc_price"],
  comp: ["tvl", "comp_price"],
  eth: ["base_fee", "eth_price", "tx_cost"],
  mkr: [],
} as const

/**
 * Protocols
 */
export type ProtocolId = keyof typeof METRIC_DECLARATIONS
export type Protocol = {
  enabled?: boolean
  id: ProtocolId
  title: string
  wip?: boolean
}

export const PROTOCOL_MAP: Record<ProtocolId, Protocol> = {
  aave: {
    id: "aave",
    title: "Aave",
  },
  btc: {
    enabled: true,
    id: "btc",
    title: "Bitcoin",
  },
  comp: {
    enabled: true,
    id: "comp",
    title: "Compound",
    wip: true,
  },
  eth: {
    enabled: true,
    id: "eth",
    title: "Ethereum",
  },
  mkr: {
    id: "mkr",
    title: "MakerDAO",
  },
} as const

export const PROTOCOL_IDS = Object.keys(PROTOCOL_MAP) as ProtocolId[]

export const PROTOCOLS = Object.values(PROTOCOL_MAP).sort((a, b) => {
  if (a.enabled === b.enabled) {
    return 0
  } else if (a.enabled && !b.enabled) {
    return -1
  } else {
    return 1
  }
})

/**
 * Metrics
 */

// Not working:
// type ValueOf<T> = T[keyof T];
// export type MetricId = ValueOf<typeof METRIC_LIST_DEFINITION>[number];
export type MetricIdForProtocol<T extends ProtocolId> = typeof METRIC_DECLARATIONS[T][number]
export type MetricId = MetricIdForProtocol<ProtocolId>

export type Variant = {
  label: string
  precision: number
  value?: string
}

export type Metric = {
  allowCompactPriceScale?: boolean
  disallowAlerts?: boolean
  disallowCandleType?: boolean
  disallowLiveMode?: boolean
  id: MetricId
  precision: number
  preferredLogScale?: boolean
  priceUnits: PriceUnit[]
  protocol: ProtocolId
  significantDigits: number[]
  timeframes: Timeframe[]
  title: string
  variants?: Variant[]
  wip?: boolean
}

// type MapType<T extends ProtocolId> = Record<
//   T,
//   Record<MetricIdsOfProtocol<T>, Metric>
// >;
type MetricsMapType = Record<ProtocolId, Partial<Record<MetricId, Metric>>>

export const METRICS_MAP: MetricsMapType = {
  aave: {},
  btc: {
    btc_price: {
      id: "btc_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "btc",
      significantDigits: [0],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "BTC price",
    },
  },
  comp: {
    comp_price: {
      id: "comp_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "comp",
      significantDigits: [2],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Comp price",
    },
    tvl: {
      allowCompactPriceScale: true,
      disallowCandleType: true,
      id: "tvl",
      precision: 1,
      priceUnits: [PriceUnit.USD],
      protocol: "comp",
      significantDigits: [0],
      timeframes: ["Hour", "Day"],
      title: "Total value locked",
      variants: [{ label: "V3 USDC - Wrapped Ether Pool", precision: 1 }],
      wip: true,
    },
  },
  eth: {
    base_fee: {
      id: "base_fee",
      precision: 1e9,
      preferredLogScale: true,
      priceUnits: [PriceUnit.GWEI],
      protocol: "eth",
      significantDigits: [2],
      timeframes: allTimeframes,
      title: "Base fee per gas",
    },
    eth_price: {
      id: "eth_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "eth",
      significantDigits: [2],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Ether price",
    },
    tx_cost: {
      id: "tx_cost",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD, PriceUnit.ETH],
      protocol: "eth",
      significantDigits: [2, 5],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Transaction cost",
      variants: [
        { label: "ETH Transfer", precision: 1e18 / 21_000 },
        { label: "ERC20 Approval", precision: 1e18 / 45_000 },
        { label: "ERC20 Transfer", precision: 1e18 / 65_000 },
        { label: "NFT Transfer", precision: 1e18 / 85_000 },
        { label: "Uniswap V2 Swap", precision: 1e18 / 150_000 },
        { label: "Uniswap V3 Swap", precision: 1e18 / 185_000 },
        { label: "L2 Deposit", precision: 1e18 / 250_000 },
      ],
    },
  },
  mkr: {},
}

export const METRICS = PROTOCOL_IDS.map((protocolId) =>
  Object.values(METRICS_MAP[protocolId] || {})
).flat()

export function getMetric(protocolId: ProtocolId, metricId: MetricId): Metric {
  const metric = METRICS_MAP[protocolId][metricId]

  if (!metric) {
    throw new Error("Metric not found.")
  }

  return metric
}

export function getMetricPrecision(metric: Metric, variantIndex: number): number {
  const { variants, precision: defaultPrecision } = metric
  return variants ? variants[variantIndex].precision : defaultPrecision
}

export function getSignificantDigits(metric: Metric, priceUnitIndex: number): number {
  return metric.significantDigits[priceUnitIndex]
}
