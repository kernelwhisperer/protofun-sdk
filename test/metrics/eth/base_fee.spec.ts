import { expect, it } from "vitest"

import query, { subscribe } from "../../../src/metrics/eth/base_fee"
import { wait } from "../../../src/utils"

it("Blocks", async () => {
  // act
  const candles = await query({
    timeframe: "Block",
    until: "1628166868",
  })
  // assert
  expect(candles).toMatchSnapshot()
})

it("Daily candles", async () => {
  // act
  const candles = await query({
    since: "1628121600",
    timeframe: "Day",
    until: "1628553600",
  })
  // assert
  expect(candles).toMatchSnapshot()
})

it("Weekly candles", async () => {
  // act
  const candles = await query({
    since: "1627862400",
    timeframe: "Week",
    until: "1630886400",
  })
  // assert
  expect(candles).toMatchSnapshot()
})

it("Subscribe", async () => {
  // arrange
  const timeframe = "Minute"
  const candles = await query({
    limit: 1,
    timeframe,
  })
  // act
  const unsubscribe = subscribe({
    onNewData: (data) => {
      candles.push(data)
      unsubscribe()
    },
    pollingInterval: 100,
    since: candles[0].timestamp,
    timeframe,
  })
  // assert
  await wait(1_000)
  expect(candles.length).to.equal(2)
})
