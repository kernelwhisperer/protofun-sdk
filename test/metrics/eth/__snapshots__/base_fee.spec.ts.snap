// Vitest Snapshot v1

exports[`Blocks 1`] = `
[
  {
    "close": "1000000000",
    "high": "1000000000",
    "low": "1000000000",
    "open": "1000000000",
    "timestamp": "1628166822",
  },
  {
    "close": "1124967822",
    "high": "1124967822",
    "low": "1124967822",
    "open": "1124967822",
    "timestamp": "1628166835",
  },
  {
    "close": "1265451421",
    "high": "1265451421",
    "low": "1265451421",
    "open": "1265451421",
    "timestamp": "1628166854",
  },
  {
    "close": "1423420054",
    "high": "1423420054",
    "low": "1423420054",
    "open": "1423420054",
    "timestamp": "1628166857",
  },
  {
    "close": "1601260312",
    "high": "1601260312",
    "low": "1601260312",
    "open": "1601260312",
    "timestamp": "1628166868",
  },
]
`;

exports[`Daily candles 1`] = `
[
  {
    "close": "31482701667",
    "high": "475851627539",
    "low": "1000000000",
    "open": "1000000000",
    "timestamp": "1628121600",
  },
  {
    "close": "42893399308",
    "high": "168558792698",
    "low": "18570858888",
    "open": "34641694922",
    "timestamp": "1628208000",
  },
  {
    "close": "70230099613",
    "high": "84047627015",
    "low": "23057420646",
    "open": "43452165186",
    "timestamp": "1628294400",
  },
  {
    "close": "29878910220",
    "high": "798484721471",
    "low": "16208986177",
    "open": "79006768622",
    "timestamp": "1628380800",
  },
  {
    "close": "37165550299",
    "high": "195220167163",
    "low": "17832066998",
    "open": "26921227286",
    "timestamp": "1628467200",
  },
  {
    "close": "38606754635",
    "high": "1866471906140",
    "low": "21950599207",
    "open": "41134634288",
    "timestamp": "1628553600",
  },
]
`;

exports[`Subscribe 1`] = `
[
  {
    "close": "9900346155",
    "high": "10048136236",
    "low": "9485515605",
    "open": "9650738006",
    "timestamp": "1695122400",
  },
]
`;

exports[`Weekly candles 1`] = `
[
  {
    "close": "29878910220",
    "high": "798484721471",
    "low": "1000000000",
    "open": "1000000000",
    "timestamp": "1627862400",
  },
  {
    "close": "38780276579",
    "high": "1866471906140",
    "low": "17290998564",
    "open": "26921227286",
    "timestamp": "1628467200",
  },
  {
    "close": "23827814264",
    "high": "704414537754",
    "low": "11689438352",
    "open": "43538780365",
    "timestamp": "1629072000",
  },
  {
    "close": "70638856142",
    "high": "2889181363031",
    "low": "17935408442",
    "open": "26804163353",
    "timestamp": "1629676800",
  },
  {
    "close": "84176648356",
    "high": "2444011563222",
    "low": "37833636195",
    "open": "79460053424",
    "timestamp": "1630281600",
  },
  {
    "close": "41703237368",
    "high": "3941920362218",
    "low": "28689995635",
    "open": "81987919992",
    "timestamp": "1630886400",
  },
]
`;
